Pins in the manifest

mosteo opened the pull request:
Pins are now to be defined in the manifest:

```
[[pins]]
foo = { version = "1.3.2+bugfix" }
bar = { path = "../my/bar" }
baz = { url = "https://github.com/baz.git" } # No commit, will use current HEAD, will update on `alr update`
gru = { url = "https://gitlab.com/gru.git" commit="123456890abcdef..." } # Explicit commit, won't update
```

This PR is still incomplete, but is testable: the new pins work, and the old pins also work, but such old pins will be forgotten after any update (explicitly requested, or by editing the manifest), as the update process syncs with the pins in the manifest.

To do:
- [x] Remove old pins
- [x] Tests specific to the new format

Fixes #721

mosteo committed:
Added data structures

mosteo committed:
Pins may appear in manifest, being ignored

mosteo committed:
Loading of user pins complete

mosteo committed:
Pins are downloaded or skipped as needed

mosteo committed:
Pins are properly pruned, and info displayed

Minor testsuite tweak for a change in logging format

mosteo committed:
Added new-format pins to alire.toml for self-build

mosteo committed:
Allow selective update of pins as for regular deps

mosteo committed:
Local pins work with new manifest syntax

mosteo committed:
Fixed bug in which version pins were not used

mosteo committed:
Make version explicit key in user pin

mosteo committed:
Fix bug about confirming empty updates

mosteo committed:
Roots: Conflated dep updating into single Sync

We had two confusing Update_Dependencies and Update_And_Deploy_Dependencies
that were in practice doing almost the same. There is now a single
Sync_Dependencies.

Fabien-Chouteau commented:
> [[pins]]
> foo = { version = "1.3.2+bugfix" }
> bar = { path = "../my/bar" }
> baz = { url = "https://github.com/baz.git" } # No commit, will use current HEAD, will update on `alr update`
> gru = { url = "https://gitlab.com/gru.git" commit="123456890abcdef..." } # Explicit commit, won't update

This looks great!

onox commented on src/alire/alire-solutions-diffs.adb:
```diff
@@ -384,7 +384,10 @@ package body Alire.Solutions.Diffs is
          Table.Print (Level);
       else
-         Trace.Log (Prefix & "No changes between former an new solution.",
+         Trace.Warning (Boolean'(This.Former  = This.Latter)'Image);
```
`'Image` should be `'Img` here to avoid "prefix of "Image" attribute must be a type" error.

mosteo replied:
Strange. What compiler version is this?

onox replied:
GNAT FSF 7.5.

mosteo replied:
I realize now that those were forgotten logs to be removed, so not a real issue.

onox commented on config/alr_config.ads:
```diff
@@ -0,0 +1,4 @@
+--  Configuration for alr generated by Alire
```
Should this folder _not_ be added to git?

mosteo replied:
Fixed by #744

Fabien-Chouteau replied:
Fixed in Alire, but they still should not be checked in here.

onox commented:
I did some testing with `orka_simd`, added the following to its alire.toml file:

```toml
[[pins]]
orka_types = {path = "../orka_types"}
```

and removing alire.lock and then running:

```
alr build -XORKA_SIMD_EXT="AVX2" (says solution is incomplete)
alr update (asks me to confirm new solution)
```

It has found the folder of orka_types :+1: (and gives me the original alire.lock again)

but if I remove alire.lock and then immediately run `alr update`, I get:

```
stderr: SYSTEM.ASSERTIONS.ASSERT_FAILURE
stderr: failed precondition from alire-solutions.ads:124
stderr: raised SYSTEM.ASSERTIONS.ASSERT_FAILURE : failed precondition from alire-solutions.ads:124
```

mosteo commented:
Thanks, @onox. The PR is still undercooked so I'll keep an eye out for these issues.

mosteo committed:
Disable tests that rely on `alr pin`

For now, these cannot work as we are going to remove the ability to edit pins
via `alr with`/`alr pin`. This functionality could be reintroduced at a later
time.

mosteo committed:
More tests temporarily disabled

Most of those should be reimplemented in their manual edition alternative

mosteo committed:
Disable code for pin edition in command-line

mosteo committed:
Changes to allow pins to non-dependencies

This is purely for user comfort and will probably result in dependencies having
to be added at publish time. However, if we manage to restore command-line
pinning, we can remove that pain by adding missing dependencies at that time.

mosteo committed:
Tests in tests/pin using new pins

Some tests are not easily portable without support from `alr with --use`. Since
that should be easy to implement later, they are disabled for now and will be
enabled in a subsequent patch.

mosteo committed:
Documentation on new pins

mosteo committed:
Fix for testcase on Windows using path separators

mosteo committed:
Spelling fixes, and exclude lockfiles from check

mosteo committed:
New test for various invalid pin entries in manifest

Also the required code changes to pass the test

mosteo committed:
Fixes from code self-review

mosteo commented:
This one is ready for review. There are quite some disabled tests and some portions of unreachable code. I intend to tidy that up in a follow-up patch (since this one has grown already too large), in which at least `alr with --use` works again, since that would allows us to keep lots of tests as-is that otherwise are hard to adapt. As this will require only to append to the manifest, as is already done for `with`, it should not be too complicated (fingers crossed).

Other than that, the new pins should be ready. @onox , as you have an use case, your feedback on this one would be great.

Fabien-Chouteau commented on doc/catalog-format-spec.md:
````diff
@@ -267,7 +267,7 @@ static, i.e. they cannot depend on the context.
    ```toml
    project-files = ["my_project.gpr", "utils/utils_for_my_project.gpr"]
-
+   
````
Looks like trailing white-spaces are introduced.

mosteo replied:
Ugh. I used a new .md editor, that will be it.

Fabien-Chouteau commented on doc/catalog-format-spec.md:
```diff
@@ -508,6 +508,63 @@ static, i.e. they cannot depend on the context.
+
+It is usual to develop several interdependent crates at the same time. In this scenario, it is often impractical to rely on indexed releases which are not intended to be modified. Instead, one would prefer to use a work-in-progress version of a crate to fulfill some dependency.
+
+Alire provides *pins* to support this use case. Pins are overrides to dependencies, are intended to be used locally, and to be fulfilled by proper dependencies once a crate is ready to be published. The use of pins is based on two ideas:
```
This sentence reads strange to me. I would suggest:
`Pins override dependencies, they are intended to be used locally, and to be fulfilled by proper dependencies once a crate is ready to be published.`

Fabien-Chouteau commented on doc/catalog-format-spec.md:
````diff
@@ -508,6 +508,63 @@ static, i.e. they cannot depend on the context.
+```
+/path/to/my_crate
+├── alire.toml
+└── my_crate_test
````
I would just call the directory `tests` here.

Fabien-Chouteau commented on doc/catalog-format-spec.md:
````diff
@@ -508,6 +508,63 @@ static, i.e. they cannot depend on the context.
+[[pins]]
+my_crate = { path = ".." }  # Overridden by the latest sources
+```
+
````
You can mention dependencies that are specific for testing.

There is also a the fact that you can set crate config variables specifically for testing, but that's maybe too complex here.

Fabien-Chouteau commented on doc/catalog-format-spec.md:
```diff
@@ -508,6 +508,63 @@ static, i.e. they cannot depend on the context.
+  * `crate_name = { url = "https://my/repo.git" } # Updatable pin`
+  * `crate_name = { url = "https://my/repo.git", commit="abcdef..." } # Fixed pin`
+
+### Using pins for crate testing
```
It's great to have the example here :+1:

Fabien-Chouteau commented on src/alire/alire-dependencies-states.ads:
```diff
@@ -202,6 +211,13 @@ private
+   ---------
+
+   overriding function "=" (L, R : State) return Boolean
+   is (L.Image = R.Image);
```
Converting to String and the comparing the String sounds like a not very efficient way of comparison.
This should probably be flagged as an area for improvement.

mosteo replied:
Right. I have lost too much time trying to ascertain what's happening there before resorting to this. I'll flag it with a TODO.

raph-amiard commented:
Hello @mosteo, as I said live to Fabien, I think being able to specify a branch name in updatable pins would be a useful, or even necessary functionality in complex use cases.

Thanks for tackling this! I think it will be the tipping point for a lot of us at AdaCore to start using Alire.

Fabien-Chouteau commented:
@mosteo I tried to create a test crate for one of my project a found an issue. Here's how to reproduce:
```console
$ git clone --branch new_alr_pins https://github.com/Fabien-Chouteau/bbqueue-spark
$ cd cd bbqueue-spark/tests
$ alr run main_framed
[works fine]
$ touch alire.toml # to simulate modification of the manifest
$ alr run main_framed
# Detected changes in manifest, synchronizing workspace...
Dependencies automatically updated as follows:                                                

   New solution is incomplete.
   #  atomic  ~0.3  (missing)                                    
   ## bbqueue 0.2.0 (pin=../../bbqueue-spark,upgraded from 0.1.0)
warn: Generating possibly incomplete environment because of missing dependencies
warn: Generating possibly incomplete configuration because of missing dependencies

bbqueue.gpr:3:06: unknown project file: "atomic.gpr"
bbqueue.gpr:3:06: imported by "/tmp/bbqueue-spark/bbqueue.gpr"
bbqueue.gpr:3:06: imported by "/tmp/bbqueue-spark/tests/config/tests_config.gpr"
bbqueue.gpr:3:06: imported by "/tmp/bbqueue-spark/tests/tests.gpr"
gprbuild: "tests.gpr" processing failed
error: Command ["gprbuild", "-gnatwU", "-j0", "-p", "-P", "tests.gpr"] exited with code 4
error: Build failed

```

mosteo commented:
> Hello @mosteo, as I said live to Fabien, I think being able to specify a branch name in updatable pins would be a useful, or even necessary functionality in complex use cases.

Right, this is at my top of follow-up improvements. I'll tackle it in a follow-up PR.

mosteo committed:
Fixes suggested during code review

mosteo commented:
> @mosteo I tried to create a test crate for one of my project a found an issue. Here's how to reproduce:

I cannot reproduce it (it runs properly both times) but there's definitely something wrong because the first time around the pin is not detected. Will fix that and we'll see if it's related.

mosteo committed:
Fix for missed update when there is no lockfile

We were creating an empty lockfile, which was newer than the manifest, and thus
not triggering the expected automatic update.

mosteo committed:
Test to check pins are applied with no prior lockfile

This was a bug detected and corrected in the previous commit

mosteo commented:
Ready for another round, @Fabien-Chouteau . Let me know if this improved things for your test.

onox commented:
@mosteo Pins in wayland-ada repo seem to work, but Alire could try to simplify the relative paths of indirect dependencies, it currently just concatenates the paths. For example, in `wayland_ada_info` crate I get changes like this:

```diff
-path = "file:../wayland_ada_scanner"
+path = "file:../wayland_protocols_ada/../wayland_client_ada/../wayland_ada_scanner"
```

Not a big problem though.

onox commented:
I can run `make` just fine in wayland-ada and orka repositories, but in awt I still get some errors and exceptions.

I get a few errors when running `alr build` in the awt repository:

```
alr build -XORKA_SIMD_EXT="AVX2"
orka.gpr:1:06: unknown project file: "orka_types"
orka.gpr:1:06: imported by "/path/to/repo/orka/orka/orka.gpr"
orka.gpr:1:06: imported by "/path/to/repo/orka/orka/orka-egl.gpr"
orka.gpr:1:06: imported by "/path/to/repo/awt/awt-linux.gpr"
orka.gpr:2:06: unknown project file: "orka_simd"
orka.gpr:2:06: imported by "/path/to/repo/orka/orka/orka.gpr"
orka.gpr:2:06: imported by "/path/to/repo/orka/orka/orka-egl.gpr"
orka.gpr:2:06: imported by "/path/to/repo/awt/awt-linux.gpr"
orka.gpr:3:06: unknown project file: "orka_transforms"
orka.gpr:3:06: imported by "/path/to/repo/orka/orka/orka.gpr"
orka.gpr:3:06: imported by "/path/to/repo/orka/orka/orka-egl.gpr"
orka.gpr:3:06: imported by "/path/to/repo/awt/awt-linux.gpr"
gprbuild: "awt-linux.gpr" processing failed
```

If I go to ../orka/orka (`orka` crate in orka repo), and then run `alr build -XORKA_SIMD_EXT="AVX2"`. It builds everything just fine.

(If I run `alr update` in awt repo I get a warning: `warn: Missing variable in environment: os; 'other' expressions discarded`.)

And I get some exceptions when running `alr build -vvd` in awt repo:

```
stderr: Pin path is not a valid directory: ../orka_egl
debug: ---8<--- Exception dump begin ---8<---
debug: ALIRE.CHECKED_ERROR
debug: /path/to/repo/awt/alire.lock:
solution
states
state  3
Loading release from manifest: ../orka/orka/alire.toml
pins
orka_egl
Pin path is not a valid directory: ../orka_egl
debug: raised ALIRE.CHECKED_ERROR : alire-stored-error:1
```

```
debug: Loading release orka_egl=0.0.0
debug: Valid root found at ../orka/orka_egl
debug: Loading release wayland_client_ada=0.0.0
stderr: Pin path is not a valid directory: ../wayland_ada_scanner
debug: ---8<--- Exception dump begin ---8<---
debug: ALIRE.CHECKED_ERROR
debug: /path/to/repo/awt/alire.lock:
solution
states
state  5
Loading release from manifest: ../wayland-ada/wayland_client_ada/alire.toml
pins
wayland_ada_scanner
Pin path is not a valid directory: ../wayland_ada_scanner
debug: raised ALIRE.CHECKED_ERROR : alire-stored-error:3
```

```
debug: Loading release wayland_cursor_ada=0.0.0
stderr: Pin path is not a valid directory: ../wayland_client_ada
debug: ---8<--- Exception dump begin ---8<---
debug: ALIRE.CHECKED_ERROR
debug: /path/to/repo/awt/alire.lock:
solution
states
state  6
Loading release from manifest: ../wayland-ada/wayland_cursor_ada/alire.toml
pins
wayland_client_ada
Pin path is not a valid directory: ../wayland_client_ada
debug: raised ALIRE.CHECKED_ERROR : alire-stored-error:5
```

```
debug: Loading release wayland_egl_ada=0.0.0
stderr: Pin path is not a valid directory: ../wayland_client_ada
debug: ---8<--- Exception dump begin ---8<---
debug: ALIRE.CHECKED_ERROR
debug: /path/to/repo/awt/alire.lock:
solution
states
state  7
Loading release from manifest: ../wayland-ada/wayland_egl_ada/alire.toml
pins
wayland_client_ada
Pin path is not a valid directory: ../wayland_client_ada
debug: raised ALIRE.CHECKED_ERROR : alire-stored-error:17
```

There are many more similar and duplicate exceptions, these are just a few of them. I can commit the changes to the `alire.toml` files if that helps you with debugging.

onox commented:
I have pushed some commits with pins added to the various alire.toml files.

Also get this error in awt:

```
orka.gpr:19:04: no value defined for "os"
orka.gpr:19:20: undefined external reference "ALR_OS"
gprbuild: "awt-linux.gpr" processing failed
```

even though I have the following in `orka/alire.toml`:

```toml
[gpr-set-externals.'case(os)']
linux = { ALR_OS = "linux" }
```

mosteo commented:
> @mosteo Pins in wayland-ada repo seem to work, but Alire could try to simplify the relative paths of indirect dependencies, it currently just concatenates the paths. For example, in `wayland_ada_info` crate I get changes like this:
> 
> ```diff
> -path = "file:../wayland_ada_scanner"
> +path = "file:../wayland_protocols_ada/../wayland_client_ada/../wayland_ada_scanner"
> ```

Actually I think this is caused by the relative path simplification functions. This is in the lockfile?

mosteo commented:
> I have pushed some commits with pins added to the various alire.toml files.

Thanks, this will help me for sure.

mosteo commented:
@onox, thanks to your repos I have a clear picture of what's going wrong. I will submit the fix in a separate PR not to make this one even more complex, as it touches on recursive pin loading and the solver. I'll ping you there when ready.
